### PR TITLE
Add background job to update git repositories 

### PR DESCRIPTION
**Pull request by LogaJ**
- On receiving a git push notification we schedule a background job to update the local git repository.
- Stop updating git repositories on web requests
- Stop updating git periodically
- Remove update-git rake task

### CODE REVIEW

**Comment by LogaJ** [comment]
@FundingCircle/black-bat review pls

File: app/models/git_repository_loader.rb [hunk 1]
```diff
@@ -33,10 +33,22 @@ def load(repository_name)
     end
   end
 
+  def load(repository_name)
```
Comment (dideler) [inline app/models/git_repository_loader.rb]: I couldn't find this being used anywhere.

Comment (dideler) [inline app/models/git_repository_loader.rb]: Same for `#rugged_repository`, which is only used by this method.

Comment (andistuder) [inline app/models/git_repository_loader.rb]: I believe `load` is used here https://github.com/FundingCircle/shipment_tracker/blob/master/app/controllers/releases_controller.rb#L42

File: lib/tasks/jobs.rake [hunk 2]
```diff
@@ -56,16 +56,35 @@ namespace :jobs do
...
+        Thread.new do
+          break if @shutdown
+          loader.load_and_update(name)
+        end
+      }.each(&:join)
+
+      repos_hash_after = GitRepositoryLocation.app_remote_head_hash
+      repos_hash_changed = repos_hash_after.delete_if { |name, remote_head|
+        remote_head == repos_hash_changed[name]
+      }
```
Comment (dideler) [inline lib/tasks/jobs.rake]: I don't understand the purpose of these 3 lines of code. Any changes made to `repos_hash_changed` will be overridden right away on the next loop iteration.

File: lib/tasks/jobs.rake [hunk 3]
```diff
@@ -56,16 +56,35 @@ namespace :jobs do
...
+
+      end_time = Time.current
+      puts "[#{end_time}] Updated git in #{end_time - start_time} seconds"
```
Comment (dideler) [inline lib/tasks/jobs.rake]: Nitpick: "Updated git repos [...]"

File: spec/models/git_repository_loader_spec.rb [hunk 4]
```diff
@@ -4,28 +4,58 @@
...
+          git_repository_loader.load_and_update('non-existent-repo')
+        }.to raise_error(GitRepositoryLoader::NotFound)
+      end
+    end
+
+    context 'when the destination directory is not empty and is not a git repo' do
+      it 'raises a BadLocation exception' do
+        git_repository = git_repository_loader.load_and_update(repo_name)
+
+        srew_up_git_repository(git_repository)
```
Comment (dideler) [inline spec/models/git_repository_loader_spec.rb]: Typo: s**c**rew_*

File: spec/models/git_repository_loader_spec.rb [hunk 5]
```diff
@@ -179,4 +209,11 @@ def create_remote_commit
     test_git_repo.create_commit
     git_repository_location.update(remote_head: test_git_repo.head_oid)
   end
+
+  def srew_up_git_repository(git_repository)
+    path = git_repository.path
+    FileUtils.rm_rf(path)
+    Dir.mkdir(path)
+    File.open(File.join(path, 'foo.txt'), 'w') do |f| f.write('foo') end
```
Comment (dideler) [inline spec/models/git_repository_loader_spec.rb]: Suggest using `Tempfile` and `Dir#tmpdir`

File: spec/support/git_test_repository.rb [hunk 6]
```diff
@@ -30,13 +32,14 @@ def create_commit(author_name: 'Alice', pretend_version: nil, message: 'A new co
...
+      old_head_oid = head_oid
+
       create_rugged_commit(
-        tree_oid: oid,
-        message: message,
-        author_name: author_name,
-        time: time,
-        pretend_version: pretend_version,
+        tree_oid: oid, message: message,
+        author_name: author_name, time: time,
```
Comment (dideler) [inline spec/support/git_test_repository.rb]: Easier to read what's being passed in if it's one parameter per line.

**Comment by 0mega** [comment]
LGTM, when tests pass

**Comment by 0mega** [comment]
LGTM

**Pull request merged**